File: native/macos/Sources/ScreenFixCore/CalibrationGeometry.swift
public struct PointD: Equatable {
    public let x: Double
    public let y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public enum CalibrationPart: Equatable {
    case body
    case left
    case right
    case top
    case bottom
}

public struct CalibrationHit: Equatable {
    public let bandIndex: Int
    public let part: CalibrationPart

    public init(bandIndex: Int, part: CalibrationPart) {
        self.bandIndex = bandIndex
        self.part = part
    }
}

public enum CalibrationGeometry {
    private static let toleranceFactor = Double(sign: .plus, exponent: -48, significand: 1)

    private enum Axis {
        case horizontal
        case vertical
    }

    private enum Edge {
        case leading
        case trailing
    }

    public static func hitTest(
        point: PointD,
        frames: [RectD],
        handleSize: Double
    ) -> CalibrationHit? {
        guard point.x.isFinite, point.y.isFinite, handleSize.isFinite, handleSize >= 0 else {
            return nil
        }
        for index in frames.indices.reversed() {
            let frame = frames[index]
            guard isFinite(frame), frame.width >= 0, frame.height >= 0 else { continue }
            let withinWidth = point.x >= frame.x && point.x <= frame.right
            let withinHeight = point.y >= frame.y && point.y <= frame.bottom
            if withinWidth, abs(point.y - frame.bottom) <= handleSize {
                return CalibrationHit(bandIndex: index, part: .bottom)
            }
            if withinWidth, abs(point.y - frame.y) <= handleSize {
                return CalibrationHit(bandIndex: index, part: .top)
            }
            if withinHeight, abs(point.x - frame.right) <= handleSize {
                return CalibrationHit(bandIndex: index, part: .right)
            }
            if withinHeight, abs(point.x - frame.x) <= handleSize {
                return CalibrationHit(bandIndex: index, part: .left)
            }
        }
        for index in frames.indices.reversed() {
            let frame = frames[index]
            guard isFinite(frame), frame.width >= 0, frame.height >= 0 else { continue }
            if point.x >= frame.x, point.x <= frame.right,
               point.y >= frame.y, point.y <= frame.bottom {
                return CalibrationHit(bandIndex: index, part: .body)
            }
        }
        return nil
    }

    public static func drag(
        band: NormalizedRect,
        part: CalibrationPart,
        delta: PointD,
        displaySize: RectD,
        minimumSize: Double
    ) -> NormalizedRect {
        guard isFinite(band), isFinite(displaySize),
              displaySize.width > 0, displaySize.height > 0,
              delta.x.isFinite, delta.y.isFinite,
              minimumSize.isFinite, minimumSize >= 0 else {
            return band
        }

        var x = band.x
        var y = band.y
        var width = band.w
        var height = band.h
        switch part {
        case .body:
            x = clamp(x + delta.x / displaySize.width, minimum: 0, maximum: 1 - width)
            y = clamp(y + delta.y / displaySize.height, minimum: 0, maximum: 1 - height)
        case .left:
            let fixedRight = x + width
            x = clampEdge(
                current: x,
                delta: delta.x / displaySize.width,
                minimum: 0,
                maximum: max(0, fixedRight - minimumSize / displaySize.width)
            )
            width = fixedRight - x
        case .right:
            let right = clampEdge(
                current: x + width,
                delta: delta.x / displaySize.width,
                minimum: min(1, x + minimumSize / displaySize.width),
                maximum: 1
            )
            width = right - x
        case .top:
            let fixedBottom = y + height
            y = clampEdge(
                current: y,
                delta: delta.y / displaySize.height,
                minimum: 0,
                maximum: max(0, fixedBottom - minimumSize / displaySize.height)
            )
            height = fixedBottom - y
        case .bottom:
            let bottom = clampEdge(
                current: y + height,
                delta: delta.y / displaySize.height,
                minimum: min(1, y + minimumSize / displaySize.height),
                maximum: 1
            )
            height = bottom - y
        }
        return NormalizedRect(x: x, y: y, w: width, h: height)
    }

    public static func snap(
        rawBand: NormalizedRect,
        activeIndex: Int,
        part: CalibrationPart,
        bands: [NormalizedRect],
        displaySize: RectD,
        threshold: Double
    ) -> NormalizedRect {
        guard isNormalized(rawBand), bands.indices.contains(activeIndex),
              isFinite(displaySize), displaySize.width > 0, displaySize.height > 0,
              threshold.isFinite, threshold >= 0 else {
            return rawBand
        }

        switch part {
        case .body:
            let horizontalTargets = targets(
                axis: .horizontal,
                screenTargets: [0, 1],
                bands: bands,
                activeIndex: activeIndex
            )
            let verticalTargets = targets(
                axis: .vertical,
                screenTargets: [0, 1],
                bands: bands,
                activeIndex: activeIndex
            )
            let horizontallySnapped = snapAxis(
                rawBand,
                axis: .horizontal,
                edges: [.leading, .trailing],
                targets: horizontalTargets,
                threshold: threshold,
                displaySize: displaySize,
                part: part,
                resizeEdge: nil
            )
            return snapAxis(
                horizontallySnapped,
                axis: .vertical,
                edges: [.leading, .trailing],
                targets: verticalTargets,
                threshold: threshold,
                displaySize: displaySize,
                part: part,
                resizeEdge: nil
            )
        case .left:
            return snapResize(
                rawBand,
                axis: .horizontal,
                edge: .leading,
                screenTarget: 0,
                activeIndex: activeIndex,
                part: part,
                bands: bands,
                displaySize: displaySize,
                threshold: threshold
            )
        case .right:
            return snapResize(
                rawBand,
                axis: .horizontal,
                edge: .trailing,
                screenTarget: 1,
                activeIndex: activeIndex,
                part: part,
                bands: bands,
                displaySize: displaySize,
                threshold: threshold
            )
        case .top:
            return snapResize(
                rawBand,
                axis: .vertical,
                edge: .leading,
                screenTarget: 0,
                activeIndex: activeIndex,
                part: part,
                bands: bands,
                displaySize: displaySize,
                threshold: threshold
            )
        case .bottom:
            return snapResize(
                rawBand,
                axis: .vertical,
                edge: .trailing,
                screenTarget: 1,
                activeIndex: activeIndex,
                part: part,
                bands: bands,
                displaySize: displaySize,
                threshold: threshold
            )
        }
    }

    private static func snapResize(
        _ band: NormalizedRect,
        axis: Axis,
        edge: Edge,
        screenTarget: Double,
        activeIndex: Int,
        part: CalibrationPart,
        bands: [NormalizedRect],
        displaySize: RectD,
        threshold: Double
    ) -> NormalizedRect {
        snapAxis(
            band,
            axis: axis,
            edges: [edge],
            targets: targets(
                axis: axis,
                screenTargets: [screenTarget],
                bands: bands,
                activeIndex: activeIndex
            ),
            threshold: threshold,
            displaySize: displaySize,
            part: part,
            resizeEdge: edge
        )
    }

    private static func snapAxis(
        _ band: NormalizedRect,
        axis: Axis,
        edges: [Edge],
        targets: [Double],
        threshold: Double,
        displaySize: RectD,
        part: CalibrationPart,
        resizeEdge: Edge?
    ) -> NormalizedRect {
        let pointScale = axis == .horizontal ? displaySize.width : displaySize.height
        let tolerance = pointScale * toleranceFactor
        var best: NormalizedRect?
        var bestDistance: Double?
        for target in targets {
            for edge in edges {
                let leading = axis == .horizontal ? band.x : band.y
                let size = axis == .horizontal ? band.w : band.h
                let edgePosition = edge == .leading ? leading : leading + size
                let distance = abs(target - edgePosition) * pointScale
                let candidate = corrected(
                    band,
                    axis: axis,
                    edge: edge,
                    target: target,
                    resizeEdge: resizeEdge
                )
                if distance <= threshold + tolerance,
                   isLegalSnap(candidate, displaySize: displaySize, part: part),
                   bestDistance == nil || distance < bestDistance! - tolerance {
                    best = candidate
                    bestDistance = distance
                }
            }
        }
        return best ?? band
    }

    private static func targets(
        axis: Axis,
        screenTargets: [Double],
        bands: [NormalizedRect],
        activeIndex: Int
    ) -> [Double] {
        var result = screenTargets
        for index in bands.indices where index != activeIndex {
            let band = bands[index]
            guard isNormalized(band) else { continue }
            let leading = axis == .horizontal ? band.x : band.y
            let size = axis == .horizontal ? band.w : band.h
            result.append(leading)
            result.append(leading + size)
        }
        return result
    }

    private static func corrected(
        _ band: NormalizedRect,
        axis: Axis,
        edge: Edge,
        target: Double,
        resizeEdge: Edge?
    ) -> NormalizedRect {
        var position = axis == .horizontal ? band.x : band.y
        var size = axis == .horizontal ? band.w : band.h
        if resizeEdge == .leading {
            let fixedEnd = position + size
            position = target
            size = fixedEnd - target
        } else if resizeEdge == .trailing {
            size = target - position
        } else if edge == .leading {
            position = target
        } else {
            position = target - size
        }
        if axis == .horizontal {
            return NormalizedRect(x: position, y: band.y, w: size, h: band.h)
        }
        return NormalizedRect(x: band.x, y: position, w: band.w, h: size)
    }

    private static func isLegalSnap(
        _ band: NormalizedRect,
        displaySize: RectD,
        part: CalibrationPart
    ) -> Bool {
        guard isWithinBounds(band) else { return false }
        switch part {
        case .left, .right:
            return band.w * displaySize.width + displaySize.width * toleranceFactor >= 20
        case .top, .bottom:
            return band.h * displaySize.height + displaySize.height * toleranceFactor >= 20
        case .body:
            return true
        }
    }

    private static func clamp(_ value: Double, minimum: Double, maximum: Double) -> Double {
        max(minimum, min(value, maximum))
    }

    private static func clampEdge(
        current: Double,
        delta: Double,
        minimum: Double,
        maximum: Double
    ) -> Double {
        if current < minimum, delta <= 0 { return current }
        if current > maximum, delta >= 0 { return current }
        return clamp(current + delta, minimum: minimum, maximum: maximum)
    }

    private static func isFinite(_ rect: RectD) -> Bool {
        rect.x.isFinite && rect.y.isFinite && rect.width.isFinite && rect.height.isFinite
    }

    private static func isFinite(_ rect: NormalizedRect) -> Bool {
        rect.x.isFinite && rect.y.isFinite && rect.w.isFinite && rect.h.isFinite
    }

    private static func isNormalized(_ rect: NormalizedRect) -> Bool {
        isFinite(rect) && rect.w > 0 && rect.h > 0 && isWithinBounds(rect)
    }

    private static func isWithinBounds(_ rect: NormalizedRect) -> Bool {
        isFinite(rect) && rect.x >= 0 && rect.y >= 0 && rect.w >= 0 && rect.h >= 0
            && rect.x + rect.w <= 1 && rect.y + rect.h <= 1
    }
}

File: native/windows/src/ScreenFix.Core/Geometry/CalibrationGeometry.cs
namespace ScreenFix.Core.Geometry;

public enum DragPart
{
    Body,
    Left,
    Right,
    Top,
    Bottom,
}

public readonly record struct PointD(double X, double Y);

public readonly record struct EditorHit(int Index, DragPart Part);

public static class CalibrationGeometry
{
    private const double PointToleranceFactor = 3.552713678800501e-15;

    public static EditorHit? HitTest(
        PointD point,
        IReadOnlyList<RectD> localBands,
        double handleSize)
    {
        ArgumentNullException.ThrowIfNull(localBands);
        RequireFinite(point.X, nameof(point));
        RequireFinite(point.Y, nameof(point));
        RequirePositiveFinite(handleSize, nameof(handleSize));

        for (var index = localBands.Count - 1; index >= 0; index--)
        {
            var band = localBands[index];
            RequirePositiveRect(band, nameof(localBands));
            var withinHeight = point.Y >= band.Y && point.Y <= band.Bottom;
            var withinWidth = point.X >= band.X && point.X <= band.Right;

            if (withinWidth && Math.Abs(point.Y - band.Bottom) <= handleSize)
            {
                return new EditorHit(index, DragPart.Bottom);
            }

            if (withinWidth && Math.Abs(point.Y - band.Y) <= handleSize)
            {
                return new EditorHit(index, DragPart.Top);
            }

            if (withinHeight && Math.Abs(point.X - band.Right) <= handleSize)
            {
                return new EditorHit(index, DragPart.Right);
            }

            if (withinHeight && Math.Abs(point.X - band.X) <= handleSize)
            {
                return new EditorHit(index, DragPart.Left);
            }
        }

        for (var index = localBands.Count - 1; index >= 0; index--)
        {
            var band = localBands[index];
            if (point.X >= band.X && point.X <= band.Right &&
                point.Y >= band.Y && point.Y <= band.Bottom)
            {
                return new EditorHit(index, DragPart.Body);
            }
        }

        return null;
    }

    public static RectD DragBand(
        RectD normalizedBand,
        DragPart part,
        PointD localDelta,
        RectD fullFrame,
        double minimumSize)
    {
        RequireNormalizedRect(normalizedBand, nameof(normalizedBand));
        RequireFullFrame(fullFrame);
        RequireFinite(localDelta.X, nameof(localDelta));
        RequireFinite(localDelta.Y, nameof(localDelta));
        RequirePositiveFinite(minimumSize, nameof(minimumSize));

        var deltaX = localDelta.X / fullFrame.Width;
        var deltaY = localDelta.Y / fullFrame.Height;
        var minimumWidth = minimumSize / fullFrame.Width;
        var minimumHeight = minimumSize / fullFrame.Height;

        return part switch
        {
            DragPart.Body => MoveBody(normalizedBand, deltaX, deltaY),
            DragPart.Left => ResizeLeft(normalizedBand, deltaX, minimumWidth),
            DragPart.Right => ResizeRight(normalizedBand, deltaX, minimumWidth),
            DragPart.Top => ResizeTop(normalizedBand, deltaY, minimumHeight),
            DragPart.Bottom => ResizeBottom(normalizedBand, deltaY, minimumHeight),
            _ => throw new ArgumentOutOfRangeException(nameof(part)),
        };
    }

    public static RectD SnapBand(
        RectD rawBand,
        int activeIndex,
        DragPart part,
        IReadOnlyList<RectD> bands,
        RectD fullFrame,
        double threshold,
        double minimumSize)
    {
        ArgumentNullException.ThrowIfNull(bands);
        RequireNormalizedRect(rawBand, nameof(rawBand));
        RequireFullFrame(fullFrame);
        RequireFinite(threshold, nameof(threshold));
        RequirePositiveFinite(minimumSize, nameof(minimumSize));
        if (threshold < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(threshold));
        }

        if (activeIndex < 0 || activeIndex >= bands.Count)
        {
            throw new ArgumentOutOfRangeException(nameof(activeIndex));
        }

        return part switch
        {
            DragPart.Body => SnapBody(rawBand, activeIndex, bands, fullFrame, threshold),
            DragPart.Left => SnapAxis(
                rawBand,
                horizontal: true,
                leadingEdges: true,
                trailingEdges: false,
                BuildTargets(horizontal: true, [0], bands, activeIndex),
                threshold,
                fullFrame,
                DragPart.Left,
                minimumSize),
            DragPart.Right => SnapAxis(
                rawBand,
                horizontal: true,
                leadingEdges: false,
                trailingEdges: true,
                BuildTargets(horizontal: true, [1], bands, activeIndex),
                threshold,
                fullFrame,
                DragPart.Right,
                minimumSize),
            DragPart.Top => SnapAxis(
                rawBand,
                horizontal: false,
                leadingEdges: true,
                trailingEdges: false,
                BuildTargets(horizontal: false, [0], bands, activeIndex),
                threshold,
                fullFrame,
                DragPart.Top,
                minimumSize),
            DragPart.Bottom => SnapAxis(
                rawBand,
                horizontal: false,
                leadingEdges: false,
                trailingEdges: true,
                BuildTargets(horizontal: false, [1], bands, activeIndex),
                threshold,
                fullFrame,
                DragPart.Bottom,
                minimumSize),
            _ => throw new ArgumentOutOfRangeException(nameof(part)),
        };
    }

    private static RectD SnapBody(
        RectD rawBand,
        int activeIndex,
        IReadOnlyList<RectD> bands,
        RectD fullFrame,
        double threshold)
    {
        var result = SnapAxis(
            rawBand,
            horizontal: true,
            leadingEdges: true,
            trailingEdges: true,
            BuildTargets(horizontal: true, [0, 1], bands, activeIndex),
            threshold,
            fullFrame,
            null,
            null);

        return SnapAxis(
            result,
            horizontal: false,
            leadingEdges: true,
            trailingEdges: true,
            BuildTargets(horizontal: false, [0, 1], bands, activeIndex),
            threshold,
            fullFrame,
            null,
            null);
    }

    private static RectD SnapAxis(
        RectD rect,
        bool horizontal,
        bool leadingEdges,
        bool trailingEdges,
        IReadOnlyList<double> targets,
        double threshold,
        RectD fullFrame,
        DragPart? resizePart,
        double? minimumSize)
    {
        var pointScale = horizontal ? fullFrame.Width : fullFrame.Height;
        var tolerance = pointScale * PointToleranceFactor;
        RectD? best = null;
        double? bestDistance = null;

        foreach (var target in targets)
        {
            if (leadingEdges)
            {
                ConsiderSnap(target, trailing: false);
            }

            if (trailingEdges)
            {
                ConsiderSnap(target, trailing: true);
            }
        }

        return best ?? rect;

        void ConsiderSnap(double target, bool trailing)
        {
            var leading = horizontal ? rect.X : rect.Y;
            var size = horizontal ? rect.Width : rect.Height;
            var edge = trailing ? leading + size : leading;
            var distance = Math.Abs(target - edge) * pointScale;
            var candidate = CorrectedRect(rect, horizontal, trailing, target, resizePart);

            if (distance <= threshold + tolerance &&
                IsSnapCandidate(candidate, fullFrame, tolerance, minimumSize) &&
                (bestDistance is null || distance < bestDistance.Value - tolerance))
            {
                best = candidate;
                bestDistance = distance;
            }
        }
    }

    private static IReadOnlyList<double> BuildTargets(
        bool horizontal,
        IReadOnlyList<double> screenTargets,
        IReadOnlyList<RectD> bands,
        int activeIndex)
    {
        var targets = new List<double>(screenTargets);
        for (var index = 0; index < bands.Count; index++)
        {
            if (index == activeIndex || !IsNormalizedRect(bands[index]))
            {
                continue;
            }

            var band = bands[index];
            targets.Add(horizontal ? band.X : band.Y);
            targets.Add(horizontal ? band.Right : band.Bottom);
        }

        return targets;
    }

    private static RectD CorrectedRect(
        RectD rect,
        bool horizontal,
        bool trailing,
        double target,
        DragPart? resizePart)
    {
        if (horizontal)
        {
            if (resizePart == DragPart.Left)
            {
                return new RectD(target, rect.Y, rect.Right - target, rect.Height);
            }

            if (resizePart == DragPart.Right)
            {
                return new RectD(rect.X, rect.Y, target - rect.X, rect.Height);
            }

            return new RectD(
                trailing ? target - rect.Width : target,
                rect.Y,
                rect.Width,
                rect.Height);
        }

        if (resizePart == DragPart.Top)
        {
            return new RectD(rect.X, target, rect.Width, rect.Bottom - target);
        }

        if (resizePart == DragPart.Bottom)
        {
            return new RectD(rect.X, rect.Y, rect.Width, target - rect.Y);
        }

        return new RectD(
            rect.X,
            trailing ? target - rect.Height : target,
            rect.Width,
            rect.Height);
    }

    private static RectD MoveBody(RectD band, double deltaX, double deltaY) => new(
        Math.Clamp(band.X + deltaX, 0, 1 - band.Width),
        Math.Clamp(band.Y + deltaY, 0, 1 - band.Height),
        band.Width,
        band.Height);

    private static RectD ResizeLeft(RectD band, double delta, double minimumWidth)
    {
        var right = band.Right;
        var left = Math.Clamp(band.X + delta, 0, Math.Max(0, right - minimumWidth));
        return new RectD(left, band.Y, right - left, band.Height);
    }

    private static RectD ResizeRight(RectD band, double delta, double minimumWidth)
    {
        var right = Math.Clamp(band.Right + delta, Math.Min(1, band.X + minimumWidth), 1);
        return new RectD(band.X, band.Y, right - band.X, band.Height);
    }

    private static RectD ResizeTop(RectD band, double delta, double minimumHeight)
    {
        var bottom = band.Bottom;
        var top = Math.Clamp(band.Y + delta, 0, Math.Max(0, bottom - minimumHeight));
        return new RectD(band.X, top, band.Width, bottom - top);
    }

    private static RectD ResizeBottom(RectD band, double delta, double minimumHeight)
    {
        var bottom = Math.Clamp(band.Bottom + delta, Math.Min(1, band.Y + minimumHeight), 1);
        return new RectD(band.X, band.Y, band.Width, bottom - band.Y);
    }

    private static bool IsSnapCandidate(
        RectD rect,
        RectD fullFrame,
        double tolerance,
        double? minimumSize) =>
        IsNormalizedRect(rect) &&
        (minimumSize is null ||
            ((rect.Width * fullFrame.Width) + tolerance >= minimumSize.Value &&
             (rect.Height * fullFrame.Height) + tolerance >= minimumSize.Value));

    private static bool IsNormalizedRect(RectD rect) =>
        IsFiniteRect(rect) &&
        rect.X >= 0 &&
        rect.Y >= 0 &&
        rect.Width > 0 &&
        rect.Height > 0 &&
        rect.Right <= 1 &&
        rect.Bottom <= 1;

    private static void RequireNormalizedRect(RectD rect, string parameterName)
    {
        if (!IsNormalizedRect(rect))
        {
            throw new ArgumentOutOfRangeException(parameterName);
        }
    }

    private static void RequirePositiveRect(RectD rect, string parameterName)
    {
        if (!IsFiniteRect(rect) || rect.Width <= 0 || rect.Height <= 0)
        {
            throw new ArgumentOutOfRangeException(parameterName);
        }
    }

    private static void RequireFullFrame(RectD fullFrame)
    {
        if (!IsFiniteRect(fullFrame) || fullFrame.Width <= 0 || fullFrame.Height <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(fullFrame));
        }
    }

    private static bool IsFiniteRect(RectD rect) =>
        double.IsFinite(rect.X) &&
        double.IsFinite(rect.Y) &&
        double.IsFinite(rect.Width) &&
        double.IsFinite(rect.Height);

    private static void RequireFinite(double value, string parameterName)
    {
        if (!double.IsFinite(value))
        {
            throw new ArgumentOutOfRangeException(parameterName);
        }
    }

    private static void RequirePositiveFinite(double value, string parameterName)
    {
        RequireFinite(value, parameterName);
        if (value <= 0)
        {
            throw new ArgumentOutOfRangeException(parameterName);
        }
    }
}
